internal/cmd/shell.go

```diff
@@ -90,7 +91,113 @@ func shell(cmd *cobra.Command, cmdArgs []string) error {
 
 	envName := cmdArgs[0]
 
-	if err := runCommand(ctx, envName, command, args); err != nil {
+	// Before the command is run, ensure the workspace is on and ready to accept
+	// an ssh connection.
+	client, err := newClient(ctx)
+	if err != nil {
+		return err
+	}
+
+	env, err := findEnv(ctx, client, envName, coder.Me)
+	if err != nil {
+		return err
+	}
+
+	var requiredRebuilds []string
+
+	// check if there are required rebuilds
+	for _, r := range env.RebuildMessages {
+		if r.Required {
+			requiredRebuilds = append(requiredRebuilds, clog.Causef(r.Text))
+		}
+	}
+
+	switch {
+	// If this conditonal is true, a rebuild is **required** to make
+	// the sh command work.
+	case env.LatestStat.ContainerStatus == coder.EnvironmentOff || len(requiredRebuilds) > 0:
+		// TODO: (@emyrk) I'd like to add a --force and --verbose flags to this command,
+		//					but currently DisableFlagParsing is set to true.
+		//					To enable force/verbose, we'd have to parse the flags ourselves,
+		//					or make the user `coder sh <env> -- [args]`
+		//
+
+		var rebuildReason string // Either because it is off, or a rebuild is required
+		if env.LatestStat.ContainerStatus == coder.EnvironmentOff {
+			rebuildReason = fmt.Sprintf("Environment %q is currently not running.", env.Name)
+		} else {
+			// Print out the reasons the rebuild is required. Newlines don't work well with
+			// the prompt label, so printing it above the prompt will have to do.
+			clog.LogWarn("rebuild required", requiredRebuilds...)
+			rebuildReason = fmt.Sprintf("Environment %q requires a rebuild to work correctly.", env.Name)
+		}
+
+		confirm := promptui.Prompt{
+			Label:     rebuildReason + " Do you wish to rebuild it now? (this will take a moment)",
+			IsConfirm: true,
+		}
+		if _, err := confirm.Run(); err != nil {
```

Comment:
we should only prompt the user if they are in an interactive terminal

Reply:
Is this what you had in mind to check for that?
https://github.com/cdr/coder-cli/blob/423279755e6607dc1236066d47fd5333d25268b5/internal/cmd/shell.go#L107-L109

So these will **not** prompt:

```
$GOPATH/bin/coder sh tiny | echo
`/home/coder/go/bin/coder sh tiny`
```

Is there a specific way of executing you had in mind? I took this code from the `runCommand`, an alternative is I could use https://rosettacode.org/wiki/Check_output_device_is_a_terminal#Go. I think they do the same thing, so I kept it consistent with the other way we did it.

Reply:
Yeah that's what I had in mind. That said, point taken... this is certainly leaky behavior. Since this is more of nicety, and we can assume 90% of cases will be `coder sh [env]` without pipes or args... I think we should do the cautious thing and only show the prompt when we know for certain the output is interactive. 

Regarding `coder sh tiny | echo`, erroring out as we do now with the helpful error message seems sufficient.

Reply:
I left a comment that `--force` or `--no-prompt` would be a clear way for someone to determine behavior, but adding flags to this command is non-trivial because the flag parsing is disabled

Reply:
I am assuming running this via `cron` and other ways won't trigger the prompt too. So I can see how this is helpful, but I don't think it's obvious if I was just writing some bash scripts. Bash scripts will still trigger the prompt if you just run them `./script.sh`.

Reply:
Yes, agreed. Doesn't appear to be an easy win here.